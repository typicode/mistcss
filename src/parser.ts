import {
  compile,
  DECLARATION,
  Element,
  Middleware,
  middleware,
  COMMENT,
  RULESET,
  // @ts-ignore
  SCOPE,
  serialize,
} from 'stylis'

export interface Data {
  comment: string
  className: string // foo
  tag: string // div
  attributes: Record<string, Set<string>> // data-foo: ['bar', 'baz']
  booleanAttributes: Set<string> // data-foo, data-bar
  properties: Set<string> // --foo, --bar
}

function getComment(element: Element): string {
  // @ts-ignore
  const prev = element.siblings[element.siblings.indexOf(element) - 1] as Element | undefined
  if (prev?.type === COMMENT) {
    return (prev.children as string).trim()
  }
  return ''
}

// (div.foo) -> { tag: 'div', className: 'foo' }
function parseScopeSelector(str: string): { tag: string; className: string } {
  const [tag = '', className = ''] = str.slice(1, -1).split('.')
  return { tag, className }
}

// :scope[data-foo="bar"] -> true
// :scope[data-foo] -> true
export function isAttribute(str: string): boolean {
  return str.startsWith(':scope[data-')
}

// :scope[data-foo="bar"] -> { attribute: 'data-foo',  value: 'bar' }
// :scope[data-foo] -> { attribute: 'data-foo',  value: '' }
function parseAttribute(str: string): { attribute: string; value?: string } {
  const [attribute = '', value] = str.slice(7, -1).split('=')
  return { attribute, value: value?.slice(1, -1) }
}

function update(data: Data): Middleware {
  return function(element, _index, _children, callback) {
    switch (element.type) {
      case DECLARATION:
        // Custom properties
        if ((element.props as string).startsWith('--')) {
          data.properties.add(element.props as string)
        }
        break

      case RULESET:
        ; (element.props as string[])
          .filter(isAttribute)
          .map(parseAttribute)
          .forEach(({ attribute, value }) => {
            if (value === undefined) {
              data.booleanAttributes.add(attribute)
            } else {
              const set = (data.attributes[attribute] ??= new Set())
              set.add(value)
            }
          })
        break
    }

    Array.isArray(element.children) && serialize(element.children, callback)
  }
}

export function parse(css: string): Data[] {
  const parsed: Data[] = []

  serialize(
    compile(css),
    middleware([
      (element) => {
        // Seach for @scope
        switch (element.type) {
          case SCOPE: {
            const prop = element.props[0] // (.foo)
            if (prop === undefined) return // Not supported
            const { tag, className } = parseScopeSelector(prop)

            const data: Data = {
              comment: getComment(element),
              tag,
              className,
              attributes: {},
              booleanAttributes: new Set<string>(),
              properties: new Set<string>(),
            }
            parsed.push(data)

            // Extract data-attributes and custom properties
            serialize(
              // Loop through @scope children
              // Note: nested rules are children of @scope, not :scope
              element.children as Element[],
              middleware([update(data)]),
            )
          }
        }
      },
    ]),
  )

  return parsed
}
