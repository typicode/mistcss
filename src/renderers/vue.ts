import { attributeToCamelCase, pascalCase, propertyToCamelCase } from './_case.js'
import { Data } from '../parser.js'
import { renderTag, renderPropsInterface, hasChildren } from './_common.js'

function renderFunction(data: Data): string {
  const args = [
    ...Object.keys(data.attributes).map(attributeToCamelCase),
    ...Array.from(data.booleanAttributes).map(attributeToCamelCase),
    ...Array.from(data.properties).map(propertyToCamelCase),
    '...props',
  ].join(', ')

  return `/**
* ${data.comment}
*/
export function ${pascalCase(data.className)}({ ${args} }: Props${hasChildren(data.tag) ? ', { slots }: SetupContext' : ''}) {
  return (${renderTag(data, '{slots.default?.()}', 'class')})
}`
}

// https://docs.astro.build/en/guides/typescript/
export function render(filename: string, data: Data): string {
  return `---
// Generated by MistCSS, do not modify
import './${filename}.mist.css'

${hasChildren(data.tag) ? "import type { SetupContext } from 'vue'" : ''}
import type { JSX } from 'vue/jsx-runtime'

${renderPropsInterface(data, `JSX.IntrinsicElements['${data.tag}']`)}

${renderFunction(data)}
`
}
