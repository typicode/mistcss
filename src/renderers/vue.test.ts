import { expect, it, describe } from 'vitest'

import { Data } from '../parser.js'
import { render } from './vue.js'

describe('render', () => {
  it('renders Vue component (full)', () => {
    const data: Data = {
      comment: 'comment',
      tag: 'div',
      className: 'foo',
      attributes: {
        'data-attr': new Set(['a', 'b']),
        'data-attr-foo-bar': new Set(['foo-bar']),
      },
      booleanAttributes: new Set(['data-is-foo']),
      properties: new Set(['--prop-foo', '--prop-bar']),
    }

    const result = render('component', data)
    expect(result).toMatchSnapshot()
  })

  it('renders Vue component (minimal)', () => {
    const data: Data = {
      comment: 'comment',
      tag: 'div',
      className: 'foo',
      attributes: {},
      booleanAttributes: new Set(),
      properties: new Set(),
    }

    const result = render('component', data)
    expect(result).toMatchSnapshot()
  })

  it('renders Vue component (void element)', () => {
    const data: Data = {
      comment: 'comment',
      tag: 'hr', // hr is a void element and should not have children
      className: 'foo',
      attributes: {},
      booleanAttributes: new Set(),
      properties: new Set(),
    }

    const result = render('component', data)
    expect(result).toMatchSnapshot()
  })
})
