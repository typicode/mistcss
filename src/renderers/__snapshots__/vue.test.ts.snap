// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`render > renders Vue component (full) 1`] = `
"---
// Generated by MistCSS, do not modify
import './component.mist.css'

import type { SetupContext } from 'vue'
import type { JSX } from 'vue/jsx-runtime'

type Props = { attr?: 'a' | 'b', attrFooBar?: 'foo-bar', isFoo?: boolean, propFoo?: string, propBar?: string } & JSX.IntrinsicElements['div']

export function Foo({ attr, attrFooBar, isFoo, propFoo, propBar, ...props }: Props, { slots }: SetupContext) {
  return (<div {...props} data-attr={attr} data-attr-foo-bar={attrFooBar} data-is-foo={isFoo} style={{ '--prop-foo': propFoo, '--prop-bar': propBar }} class="foo" >{slots.default?.()}</div>)
}
"
`;

exports[`render > renders Vue component (minimal) 1`] = `
"---
// Generated by MistCSS, do not modify
import './component.mist.css'

import type { SetupContext } from 'vue'
import type { JSX } from 'vue/jsx-runtime'

type Props = {  } & JSX.IntrinsicElements['div']

export function Foo({ ...props }: Props, { slots }: SetupContext) {
  return (<div {...props} class="foo" >{slots.default?.()}</div>)
}
"
`;

exports[`render > renders Vue component (void element) 1`] = `
"---
// Generated by MistCSS, do not modify
import './component.mist.css'


import type { JSX } from 'vue/jsx-runtime'

type Props = {  } & JSX.IntrinsicElements['hr']

export function Foo({ ...props }: Props) {
  return (<hr {...props} class="foo" />)
}
"
`;
